// github-empty-commit/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.githubEmptyCommit = void 0;
const rest_1 = require("@octokit/rest");
const github = new rest_1.Octokit({
    auth: process.env.GITHUB_TOKEN || process.env.GH_TOKEN,
});
const getReferenceCommit = async (opts) => {
    var _a;
    const { data } = await github.git.getRef({
        owner: opts.owner,
        repo: opts.repo,
        ref: (_a = opts.ref) !== null && _a !== void 0 ? _a : 'heads/master',
    });
    return data.object;
};
const getCommitData = async (opts) => {
    var _a;
    const { data } = await github.repos.getCommit({
        owner: opts.owner,
        repo: opts.repo,
        ref: (_a = opts.ref) !== null && _a !== void 0 ? _a : 'heads/master',
    });
    return data.commit.tree.sha;
};
const createCommit = async (opts, treeSha, referenceSha) => {
    var _a;
    const { data } = await github.git.createCommit({
        owner: opts.owner,
        repo: opts.repo,
        message: (_a = opts.commitMessage) !== null && _a !== void 0 ? _a : 'Empty Commit ⛳',
        tree: treeSha,
        parents: [referenceSha],
    });
    return data.sha;
};
const updateReference = async (opts, newSha) => {
    var _a;
    const { data } = await github.git.updateRef({
        owner: opts.owner,
        repo: opts.repo,
        ref: (_a = opts.ref) !== null && _a !== void 0 ? _a : 'heads/master',
        sha: newSha,
        force: true,
    });
    return { success: true, data };
};
async function githubEmptyCommit(opts) {
    const { sha } = await getReferenceCommit(opts);
    const treeSha = await getCommitData(opts);
    const commitSha = await createCommit(opts, treeSha, sha);
    const { success } = await updateReference(opts, commitSha);
    return success;
}
exports.githubEmptyCommit = githubEmptyCommit;
