// github-empty-commit/src/index.ts
import { Octokit, RestEndpointMethodTypes } from '@octokit/rest'

const github = new Octokit({
  auth: process.env.GITHUB_TOKEN || process.env.GH_TOKEN,
})

interface IOptions {
  readonly owner: string
  readonly repo: string
  readonly ref?: string
  readonly commitMessage?: string
}

const getReferenceCommit = async (opts: IOptions) => {
  const { data } = await github.git.getRef({
    owner: opts.owner,
    repo: opts.repo,
    ref: opts.ref ?? 'heads/master',
  })
  return data.object
}

const getCommitData = async (opts: IOptions) => {
  const { data } = await github.repos.getCommit({
    owner: opts.owner,
    repo: opts.repo,
    ref: opts.ref ?? 'heads/master',
  })
  return data.commit.tree.sha
}

const createCommit = async (
  opts: IOptions,
  treeSha: string,
  referenceSha: string,
) => {
  const { data } = await github.git.createCommit({
    owner: opts.owner,
    repo: opts.repo,
    message: opts.commitMessage ?? 'Empty Commit ⛳',
    tree: treeSha,
    parents: [referenceSha],
  })
  return data.sha
}

const updateReference = async (
  opts: IOptions,
  newSha: string,
): Promise<{
  success: boolean
  data: RestEndpointMethodTypes['git']['updateRef']['response']['data']
}> => {
  const { data } = await github.git.updateRef({
    owner: opts.owner,
    repo: opts.repo,
    ref: opts.ref ?? 'heads/master',
    sha: newSha,
    force: true,
  })
  return { success: true, data }
}

export async function githubEmptyCommit(opts: IOptions): Promise<boolean> {
  const { sha } = await getReferenceCommit(opts)
  const treeSha = await getCommitData(opts)
  const commitSha = await createCommit(opts, treeSha, sha)
  const { success } = await updateReference(opts, commitSha)
  return success
}
